### PR TITLE
Strike Command

### PR DESCRIPTION
**Pull request by Xenmai**
Also fixed and improved the Heal command

### CODE REVIEW

File: src/main/java/com/denizenscript/denizen2sponge/Denizen2Sponge.java [hunk 1]
```diff
@@ -265,6 +285,12 @@ public void onServerStart(GamePreInitializationEvent event) {
         // TODO: Config option -> readyToSpamEvents = true;
     }
 
+    @Listener
+    public void onServerStarted(GamePostInitializationEvent event) {
+        // Build the already loaded advancement scripts
+        AdvancementScript.buildAll();
+    }
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizen2sponge/Denizen2Sponge.java]: Why is this here, not just at the end of the method above? There's a section of calls only a few lines up from here relating to initializing/loading/registering data (flags, etc.) that it might go with, or just at the actual end of the method block.

File: src/main/java/com/denizenscript/denizen2sponge/Denizen2SpongeImplementation.java [hunk 2]
```diff
@@ -28,6 +32,11 @@ public void midLoad() {
 
     @Override
     public void reload() {
+        if (!AdvancementScript.oldAdvancementScripts.equals(AdvancementScript.currentAdvancementScripts.keySet())) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizen2sponge/Denizen2SpongeImplementation.java]: Is this well-tested? I get nervous about things like `HashSet#equals` as they rely on the Java libs doing things reasonably and stably.

File: src/main/java/com/denizenscript/denizen2sponge/Denizen2SpongeImplementation.java [hunk 3]
```diff
@@ -28,6 +32,11 @@ public void midLoad() {
 
     @Override
     public void reload() {
+        if (!AdvancementScript.oldAdvancementScripts.equals(AdvancementScript.currentAdvancementScripts.keySet())) {
+            Debug.info("Advancement scripts have changed, but won't have any effect. " +
+                    "Restart the server for the new advancements to be registered!");
+        }
+        AdvancementScript.buildAll();
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizen2sponge/Denizen2SpongeImplementation.java]: Should this `buildAll` call be here?

File: src/main/java/com/denizenscript/denizen2sponge/commands/entity/AddAITaskCommand.java [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+                case "range":
+                    if (!entry.namedArgs.containsKey("distance") || !entry.namedArgs.containsKey("delay")
+                            || !entry.namedArgs.containsKey("speed")) {
+                        queue.handleError(entry,
+                                "Tasks of type range require 'distance', 'delay' and 'speed' arguments!");
+                    }
+                    float radius = (float) NumberTag.getFor(queue.error,
+                            entry.getNamedArgumentObject(queue, "distance")).getInternal();
+                    int delay = (int) DurationTag.getFor(queue.error,
+                            entry.getNamedArgumentObject(queue, "delay")).getInternal() * 20;
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizen2sponge/commands/entity/AddAITaskCommand.java]: ^ that's wrong. Imagine the duration is half a second. You'd calculate `(int) 0.5 * 20` which steps to `0 * 20` and then just gives zero ticks.

File: src/main/java/com/denizenscript/denizen2sponge/commands/entity/RemoveAITasksCommand.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+import org.spongepowered.api.entity.ai.task.builtin.creature.AvoidEntityAITask;
+import org.spongepowered.api.entity.ai.task.builtin.creature.RangeAgentAITask;
+import org.spongepowered.api.entity.ai.task.builtin.creature.WanderAITask;
+import org.spongepowered.api.entity.ai.task.builtin.creature.horse.RunAroundLikeCrazyAITask;
+import org.spongepowered.api.entity.ai.task.builtin.creature.target.FindNearestAttackableTargetAITask;
+import org.spongepowered.api.entity.living.Agent;
+
+import java.util.Optional;
+
+public class RemoveAITasksCommand extends AbstractCommand {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizen2sponge/commands/entity/RemoveAITasksCommand.java]: You write this with command both with an `s` and without it... differently throughout the code. Choose one and make sure it's consistent throughout all references.

File: src/main/java/com/denizenscript/denizen2sponge/commands/player/TakeCommand.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    @Override
+    public void execute(CommandQueue queue, CommandEntry entry) {
+        PlayerTag player = PlayerTag.getFor(queue.error, entry.getArgumentObject(queue, 0));
+        ItemTag item = ItemTag.getFor(queue.error, entry.getArgumentObject(queue, 1));
+        if (queue.shouldShowGood()) {
+            queue.outGood("Attempting to take " + ColorSet.emphasis + item.debug() + ColorSet.good
+                    + " from " + ColorSet.emphasis + player.debug() + ColorSet.good + "!");
+        }
+        Optional<ItemStack> result = player.getInternal().getInventory()
+                .query(QueryOperationTypes.ITEM_TYPE.of(item.getInternal().getType())).poll(item.getInternal().getQuantity());
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizen2sponge/commands/player/TakeCommand.java]: A loose take-by-type is an interesting default. This setup needs a lot of expansion to work well for users.

File: src/main/java/com/denizenscript/denizen2sponge/spongescripts/AdvancementScript.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+    public static Set<String> oldAdvancementScripts = new HashSet<>();
+
+    // <--[explanation]
+    // @Since 0.4.0
+    // @Name Advancement Scripts
+    // @Group Script Types
+    // @Description
+    // An advancement script is a script that adds new advancements to the server.
+    // It is identified with the type "advancement".
+    // Expected keys: id (string), name (string, optional), criteria (map), display
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizen2sponge/spongescripts/AdvancementScript.java]: I ... have no idea why I wrote `string` in the other script type but that shouldn't be in the meta. All type specifiers should be using tag types that scripters would use, not Java type names >.> ... could you switch this one around and the game command script meta too?

File: src/main/java/com/denizenscript/denizen2sponge/spongescripts/GameCommandScript.java [hunk 8]
```diff
@@ -48,8 +53,9 @@ public static void clear() {
     // An in-game command script is a script that handles in-game commands.
     // It is simply identified with the type "command".
     // Expected keys: name (list or string), description (string), permission (string, optional),
-    // Valid contexts: source (player/block/server), player (PlayerTag, if source=player), location (LocationTag, if source=block)
-    // arguments (ListTag), raw_arguments (TextTag)
+    // script (commands). Valid contexts: source (player/block/minecart/server), player (PlayerTag,
+    // if source=player), location (LocationTag, if source=block), entity (EntityTag, if source=minecart),
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizen2sponge/spongescripts/GameCommandScript.java]: This line split is really awkwardly
placed.

File: src/main/java/com/denizenscript/denizen2sponge/tags/handlers/TextsTagBase.java [hunk 9]
```diff
@@ -67,45 +71,32 @@ public String getName() {
         // text:hello -> the base text will be 'hello'.
         // color:blue -> the color will be blue.
         // style:italics|bold -> the style will be bold-italic. Also allowed: obfuscated, reset, underline, strike.
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizen2sponge/tags/handlers/TextsTagBase.java]: Due to your change below I think `strike` became `strikethrough` and `italics` became `italic`

File: src/main/java/com/denizenscript/denizen2sponge/tags/objects/InventoryTag.java [hunk 10]
```diff
@@ -99,14 +102,14 @@ public Inventory getInternal() {
         // -->
         handlers.put("name", (dat, obj) -> new TextTag(((InventoryTag) obj).internal.getName().get(Locale.ENGLISH)));
         // @Name InventoryTag.result
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizen2sponge/tags/objects/InventoryTag.java]: I uh... just noticed that every single tag meta in this file is wrong. It's missing the top two lines. (The `<--[tag]` and `@Since`...)

File: src/main/java/com/denizenscript/denizen2sponge/Denizen2SpongeImplementation.java (hunk 2 above)
Comment (Xenmai) [reply src/main/java/com/denizenscript/denizen2sponge/Denizen2SpongeImplementation.java]: Yeah. SInce it's a "set" and not and "ordered list" it seems to work as expected every time I tested.

File: src/main/java/com/denizenscript/denizen2sponge/Denizen2SpongeImplementation.java (hunk 3 above)
Comment (Xenmai) [reply src/main/java/com/denizenscript/denizen2sponge/Denizen2SpongeImplementation.java]: Removed, right?

File: src/main/java/com/denizenscript/denizen2sponge/Denizen2Sponge.java (hunk 1 above)
Comment (Xenmai) [reply src/main/java/com/denizenscript/denizen2sponge/Denizen2Sponge.java]: Moved it there, will test soon.

File: src/main/java/com/denizenscript/denizen2sponge/commands/entity/AddAITaskCommand.java (hunk 4 above)
Comment (Xenmai) [reply src/main/java/com/denizenscript/denizen2sponge/commands/entity/AddAITaskCommand.java]: Fixed by grouping the operation with braces. Thanks for pointing it out.

File: src/main/java/com/denizenscript/denizen2sponge/spongescripts/AdvancementScript.java (hunk 7 above)
Comment (Xenmai) [reply src/main/java/com/denizenscript/denizen2sponge/spongescripts/AdvancementScript.java]: Changed it to actual D2 Tag types.

File: src/main/java/com/denizenscript/denizen2sponge/spongescripts/GameCommandScript.java (hunk 8 above)
Comment (Xenmai) [reply src/main/java/com/denizenscript/denizen2sponge/spongescripts/GameCommandScript.java]: True, fixed now.

File: src/main/java/com/denizenscript/denizen2sponge/tags/handlers/TextsTagBase.java (hunk 9 above)
Comment (Xenmai) [reply src/main/java/com/denizenscript/denizen2sponge/tags/handlers/TextsTagBase.java]: Correct. Changed it now. Also I've seen that old meta uses single quotes here and there. Is that fine, or should I change them to double quotes?

File: src/main/java/com/denizenscript/denizen2sponge/tags/objects/InventoryTag.java (hunk 10 above)
Comment (Xenmai) [reply src/main/java/com/denizenscript/denizen2sponge/tags/objects/InventoryTag.java]: Nice catch. Added the missing lines.

File: src/main/java/com/denizenscript/denizen2sponge/commands/player/CooldownCommand.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        else {
+            passed = player.getOnline(queue.error).getCooldownTracker().resetCooldown(item.getInternal());
+            if (queue.shouldShowGood()) {
+                queue.outGood("Attempting to reset cooldown of item type '" + ColorSet.emphasis + item.debug()
+                        + ColorSet.good + "' for player '" + ColorSet.emphasis + player.debug() + ColorSet.good + "'!");
+            }
+        }
+        if (passed) {
+            if (queue.shouldShowGood()) {
+                queue.outGood("Cooldown operation passed!");
```
[the remark this answers is not in the document]
Comment (Xenmai) [reply src/main/java/com/denizenscript/denizen2sponge/commands/player/CooldownCommand.java]: I've not seen the operation fail, so best approach might be to just remove that "passed/failed" output, right?

File: src/main/java/com/denizenscript/denizen2sponge/commands/player/TakeCommand.java (hunk 6 above)
Comment (Xenmai) [reply src/main/java/com/denizenscript/denizen2sponge/commands/player/TakeCommand.java]: I changed it so it takes an "operation" named argument. Available operations are "item_exact", "item" (exact item ignoring quantity) and "item_type". Should offer more possibilities.

File: src/main/java/com/denizenscript/denizen2sponge/Denizen2Sponge.java [hunk 12]
```diff
@@ -258,6 +277,8 @@ public void onServerStart(GamePreInitializationEvent event) {
         Sponge.getEventManager().post(new Denizen2SpongeLoadingEvent(getGenericCause()));
         // Load Denizen2
         Denizen2Core.start();
+        // Build loaded advamcements
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizen2sponge/Denizen2Sponge.java]: admamcememts.
(Spelling)

File: src/main/java/com/denizenscript/denizen2sponge/commands/entity/VanishCommand.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    // @Since 0.4.0
+    // @Name vanish
+    // @Arguments <entity> <state>
+    // @Short changes the vanish state of an entity.
+    // @Updated 2018/02/11
+    // @Group Entity
+    // @Minimum 2
+    // @Maximum 2
+    // @Named ignore_collisions (BooleanTag) Sets whether this vanished entity can collide with other entities.
+    // @Named untargeteable (BooleanTag) Sets whether this vanished entity can be selected as a target.
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/denizen2sponge/commands/entity/VanishCommand.java]: `untargeteable` that's not spelled right.

Comment (Xenmai) [reply src/main/java/com/denizenscript/denizen2sponge/commands/entity/VanishCommand.java]: "untargetable", right? Or should I change it to "prevents_targeting"?
